modules/dlm/src/main/java/org/elasticsearch/dlm/DataLifecycleService.java

```diff
@@ -0,0 +1,321 @@
+        if (rolloverInfo != null) {
+            return TimeValue.timeValueMillis(rolloverInfo.getTime());
+        } else {
+            return TimeValue.timeValueMillis(index.getCreationDate());
+        }
+    }
+
+    /**
+     * This is quite a shallow method but the purpose of its existence is to have only one place to modify once we
+     * introduce the index.lifecycle.prefer_ilm setting. Once the prefer_ilm setting exists the method will also
+     * make more sense as it will encapsulate a bit more logic.
+     */
+    private static boolean isManagedByDLM(DataStream parentDataStream, IndexMetadata indexMetadata) {
+        return indexMetadata.getLifecyclePolicyName() == null && parentDataStream.getLifecycle() != null;
+    }
+
+    private RolloverRequest getDefaultRolloverRequest(String dataStream) {
+        RolloverRequest rolloverRequest = new RolloverRequest(dataStream, null).masterNodeTimeout(TimeValue.MAX_VALUE);
+
+        // TODO get rollover from cluster setting once we have it
+        rolloverRequest.addMaxIndexAgeCondition(TimeValue.timeValueDays(30));
+        rolloverRequest.addMaxPrimaryShardSizeCondition(ByteSizeValue.ofGb(50));
+        rolloverRequest.addMaxPrimaryShardDocsCondition(200_000_000);
+        // don't rollover an empty index
+        rolloverRequest.addMinIndexDocsCondition(1);
+        return rolloverRequest;
+    }
+
+    private void onBecomeMaster() {
+        maybeScheduleJob();
+    }
+
+    private void updatePollInterval(TimeValue newInterval) {
+        this.pollInterval = newInterval;
+        maybeScheduleJob();
+    }
+
+    private void cancelJob() {
+        if (scheduler.get() != null) {
+            scheduler.get().remove(DATA_LIFECYCLE_JOB_NAME);
+            scheduledJob = null;
+        }
+    }
+
+    private boolean isClusterServiceStoppedOrClosed() {
+        final Lifecycle.State state = clusterService.lifecycleState();
+        return state == Lifecycle.State.STOPPED || state == Lifecycle.State.CLOSED;
+    }
+
+    private void maybeScheduleJob() {
+        if (this.isMaster) {
+            if (scheduler.get() == null) {
+                // don't create scheduler if the node is shutting down
+                if (isClusterServiceStoppedOrClosed() == false) {
+                    scheduler.set(new SchedulerEngine(settings, clock));
+                    scheduler.get().register(this);
+                }
+            }
+
+            // scheduler could be null if the node might be shutting down
+            if (scheduler.get() != null) {
+                scheduledJob = new SchedulerEngine.Job(DATA_LIFECYCLE_JOB_NAME, new TimeValueSchedule(pollInterval));
+                scheduler.get().add(scheduledJob);
+            }
+        }
+    }
+
+    // package visibility for testing
+    void setDefaultRolloverRequestSupplier(Function<String, RolloverRequest> defaultRolloverRequestSupplier) {
+        this.defaultRolloverRequestSupplier = defaultRolloverRequestSupplier;
+    }
```

Comment:
We have this to make testing easier right? I would like to get to the point where we parsing out the real rollover request from the configuration and then we can remove this entirely, what do you think?

Reply:
Absolutely. Mary is working on introducing the default rollover setting. This will go away once that's merged.